### PR TITLE
feat: add sgid oauth flow

### PR DESCRIPTION
**Pull request by orangekjc**
## Problem

_What problem are you trying to solve? What issue does this close?_
- Added draft oAuth flow for telegram verification.

Closes [insert issue #]

## Solution

_How did you solve the problem?_

**Features**:

- Added `/auth/sgid/auth-url` route to generate sgid url for authentication.
- Added `/auth/sgid/callback` route for sgid redirect.

**Improvements**:

- Details ...

**Bug Fixes**:

- Details ...

## Before & After Screenshots

**BEFORE**:
[insert screenshot here]

**AFTER**:
[insert screenshot here]

## Tests

_What tests should be run to confirm functionality?_

## Deploy Notes

_Notes regarding deployment of the contained body of work. These should note any
new dependencies, new scripts, etc._

**New environment variables**:
For initialising sgid client. 
- SGID_CLIENT_ID: The client ID provided during client registration
- SGID_CLIENT_SECRET: The client ID provided during client registration
- SGID_CLIENT_PRIVATE_KEY: The client ID provided during client registration 

**New scripts**:

- `script` : script details

**New dependencies**:

- @nestjs/config: Environment variable management
- @opengovsg/sgid-client: sgid client sdk
- class-transformer: For data transformation
- class-validator: For data validation
- cookie-parser: Enable cookies access

**New dev dependencies**:
- @types/cookie-parser: Enable typing for cookie-parser

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    const rawPoDetails =
+      userInfo.data[SGID_SCOPE_PUBLIC_OFFICER_DETAILS] ?? null;
+    const poDetails: PublicOfficerDetails[] = JSON.parse(rawPoDetails);
+
+    if (!poDetails || poDetails.length === 0) {
+      return 'You are not authorised to join this channel.';
+    } else {
+      // TODO: Send invite code via telegram bot.
+      return 'You have been verified, the invite link has been sent via our telegram bot.';
+    }
```
Comment (LoneRifle) [inline src/auth/auth.service.ts:85]: I think we should try to make this as general as possible. To that end, I have a couple of points:

- We do not know for sure that the bot will be used as an automated admin for Telegram groups. The bot could possibly be used for other purposes too, eg, allowing authenticated officers to transact with the bot itself
- To honour Separation of Concerns, perhaps the return value of this method should be a Boolean, ie, if the officer has been authenticated. You might opt for an enum to express several states, ie, not authenticated, authenticated but not a public officer, or authenticated and found to be a public officer. The controller then assumes responsibility for what to do next.

**Review by LoneRifle — CHANGES_REQUESTED** [review]
lgtm otherwise

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    const authStatus = await this.authService.verifyUserFromAuthCode({
+      code,
+      nonce: cookieInstance.nonce,
+      codeVerifier: cookieInstance.codeVerifier,
+    });
+
+    // TODO: Insert custom flow depending on the authentication status.
+    return SgidAuthStatus[authStatus];
```
Comment (LoneRifle) [inline src/auth/auth.controller.ts:76]: Hmm. Would be nice if we could 302 redirect it back to the bot at `https://t.me/<botname>` (like how the CPF app would open up Singpass login in an in-app browser view, then prompt you to go back to the app). The bot could then send a message depending on the auth outcome, perhaps invoked from this controller.

If you want a demonstration of how this works in the CPF app come find me later.

Also, why are you using an enum to index into the same enum?

Comment (orangekjc) [reply src/auth/auth.controller.ts:76]: Yeap, I'll give it a go!

sorry I was thinking to reflect the Enum text on FE during development 😅

**Review by LoneRifle — APPROVED** [review]
lgtm for now. TODO:
- Stashing a cookie wouldn't quite help since the session is almost completely on Telegram, not our service. We need to figure a way to somehow bind their identifiers across Telegram and sgID. This could be done either with a database, or if possible, on Telegram itself
- We should rework the redirect so that the actual URL is driven by the configured bot name at runtime. I think once the PR to add telegraf happens we can tie the two together

**Pull request merged**